File: exercise/week04/c#/Construct/BackdoorPassUpdater.cs
namespace Construct;

public class BackdoorPassUpdater : NormalArtefactUpdater, IArtefactUpdater
{
    void IArtefactUpdater.Update(Artefact artefact)
    {
        Increase(artefact, 1);
        if (artefact.TimeToLive < 11) Increase(artefact, 1);
        if (artefact.TimeToLive < 6) Increase(artefact, 1);

        artefact.TimeToLive--;

        if (artefact.TimeToLive < 0)
            artefact.Integrity = 0;
    }
}

File: exercise/week04/ts/src/construct/backdoorPassUpdater.ts
import { Artefact } from "./artefact";
import { ArtefactUpdater, BaseArtefactUpdater } from "./artefactUpdater";

export class BackdoorPassUpdater extends BaseArtefactUpdater implements ArtefactUpdater {
    update(artefact: Artefact): void {
        this.increase(artefact, 1);
        if (artefact.timeToLive < 11) this.increase(artefact, 1);
        if (artefact.timeToLive < 6) this.increase(artefact, 1);

        artefact.timeToLive--;

        if (artefact.timeToLive < 0) {
            artefact.integrity = 0;
        }
    }
}